docs/config.md
---
sidebar_position: 7
# SPDX-FileCopyrightText: 2023 Steffen Vogel <post@steffenvogel.de>
# SPDX-License-Identifier: Apache-2.0
---

# Configuration

This page describes the ways of configuring the cunīcu daemon (`cunicu daemon`).

## Command Line Flags

Basic options of `cunicu daemon` can be configured by passing command line arguments.
A full overview is available in its [manpage](./usage/md/cunicu_daemon.md).

## Configuration File

For more advanced setups, a configuration file can be used for a persistent configuration:

Please have a look at the [example configuration file](./config-reference.md) for a full reference of all available settings.

## Environment Variables

All the settings from the configuration file can also be passed via environment variables by following the following rules:

-   Convert the setting name to uppercase
-   Prefixing the setting name with `CUNICU_`
-   Nested settings are separated by underscores

**Example:** The setting `ice.max_binding_requests` can be set by the environment variable `CUNICU_ICE_MAX_BINDING_REQUESTS`

:::note
Setting lists such as `ice.urls` or `backends` can currently not be set via environment variables.
:::

## At Runtime

cunīcu's configuration can also be updated at runtime, elevating the need to restart the daemon to avoid interruption of connectivity.

Please have a look at the [`cunicu config`](./usage/md/cunicu_config.md) commands.

## DNS Auto-configuration

cunīcu als supports retrieving parts of the configuration via DNS lookups.
This is useful for corporate environments in which a fleet of cunīcu daemon need to be configured centrally.

In this case `cunicu daemon` is started one or more `--domain example.com` parameters to look for the following DNS records to obtain its configuration.

STUN and TURN servers used for ICE are retrieved by SVR lookups and other cunīcu settings are retrieved via `SRV` and `TXT` lookups: 

```text
_stun._udp.example.com.  3600 IN SRV 10 0 3478 stun.example.com.
_stuns._tcp.example.com. 3600 IN SRV 10 0 3478 stun.example.com.
_turn._udp.example.com.  3600 IN SRV 10 0 3478 turn.example.com.
_turn._tcp.example.com.  3600 IN SRV 10 0 3478 turn.example.com.
_turns._tcp.example.com. 3600 IN SRV 10 0 5349 turn.example.com.

example.com.             3600 IN TXT "cunicu-config=https://example.com/cunicu.yaml"
example.com.             3600 IN TXT "cunicu-backend=grpc://signal.example.com:443"
example.com.             3600 IN TXT "cunicu-community=my-community-password"
example.com.             3600 IN TXT "cunicu-ice-username=user1"
example.com.             3600 IN TXT "cunicu-ice-password=pass1"
```

:::note
The `cunicu-backend` and `cunicu-config` TXT records can be provided multiple times. Others not.
:::

## Remote Configuration Files

When `cunicu daemon` can be started with `--config` options pointing to HTTPS URIs:

```bash
cunicu daemon --config http://example.com/cunicu.yaml
```

cunīcu will download all configuration files in the order they are specified on the command line and merge them subsequently.

This feature can be combined with the DNS auto-configuration method by providing a TXT record pointing to the configuration file:

```text
example.com. 3600 IN TXT "cunicu-config=https://example.com/cunicu.yaml"
```

:::note
Remote configuration files must be fetched via HTTPS if they are not hosted locally and required a trusted server certificate.
:::

## Auto-reload

cunīcu watches local and remote files as well as the DNS configuration for changes and automatically reloads its configuration from them whenever a change has been detected.

For local files the change is detected by [inotify(7)](https://man7.org/linux/man-pages/man7/inotify.7.html).
For remote sources, cunīcu periodically checks the `Last-Modified` and `Etag` headers in case of HTTP files or the DNS zone's [SOA serial number](https://en.wikipedia.org/wiki/SOA_record#Structure) to detect changes without request the full remote source.

:::note
Configuration file distributed via `conicu-config` DNS TXT record are not yet monitored for changes.
:::


docs/Configuration.md
# Configuration

This page describes the ways of configuring the cunicu daemon (`cunicu daemon`).

## Command Line Flags

The `cunicu daemon` can almost fully be configured by passing command line arguments.
A full overview is available in its [manpage](./usage/md/cunicu_daemon.md).

## Configuration File

Alternatively a configuration file can be used for a persistent configuration:

```yaml title="cunicu.yaml"
watch_interval: 1s

backends:
- grpc://localhost:8080?insecure=true&skip_verify=true
- k8s:///path/to/your/kubeconfig.yaml?namespace=default

# WireGuard settings
wireguard:  
  # Create WireGuard interfaces using bundled wireguard-go Userspace implementation
  # This will be the default if there is no WireGuard kernel module present.
  userspace: false

  # Ignore WireGuard interface which do not match this regular expression
  interface_filter: .*

  # A list of WireGuard interfaces which should be configured
  interfaces:
  - wg-vpn

  # Port range for ListenPort setting of newly created WireGuard interfaces
  # cunīcu will select the first available port in this range.
  port:
    min: 52820
    max: 65535

# Control socket settings
socket:
  path: /var/run/cunicu.sock

  # Start of cunīcu daemon will block until its unblocked via the control socket
  # Mostly useful for testing automation
  wait: false

# Synchronize WireGuard interface configurations with wg(8) config-files.
config_sync:
  enabled: false
  
  # Directory where Wireguard configuration files are located.
  # We expect the same format as used by wg(8) and wg-quick(8).
  # Filenames must match the interface name with a '.conf' suffix.
  path: /etc/wireguard

  # Watch the configuration files for changes and apply them accordingly.
  watch: false
  
# Synchronize WireGuard AllowedIPs with Kernel routing table
route_sync:
  enabled: true

  table: main

# Discovery of other WireGuard peers
peer_disc:
  enabled: true

  # A list of WireGuard public keys which are accepted peers
  whitelist:
  - coNsGPwVPdpahc8U+dbbWGzTAdCd6+1BvPIYg10wDCI=
  - AOZzBaNsoV7P8vo0D5UmuIJUQ7AjMbHbGt2EA8eAuEc=

  # A passphrase shared among all peers of the same community
  community: "some-common-password"

# Discovery of WireGuard endpoint addressesendpoint_disc:
  enabled: true

  # Interactive Connectivity Establishment parameters
  ice:
    # A list of STUN and TURN servers used by ICE
    urls:
    - stun:stun.l.google.com:19302

    # Credentials for STUN/TURN servers configured above
    username: ""
    password: ""

    # Allow connections to STUNS/TURNS servers for which
    # we cant validate their TLS certificates
    insecure_skip_verify: false

    # Limit available network and candidate types
    network_types: [udp4, udp6, tcp4, tcp6]
    candidate_types: [host, srflx, prflx ,relay]

    # Regular expression whitelist of interfaces which are used to gather ICE candidates.
    interface_filter: .*

    # Lite agents do not perform connectivity check and only provide host candidates.
    lite: false

    # Attempt to find candidates via mDNS discovery
    mdns: false

    # Sets the max amount of binding requests the agent will send over a candidate pair for validation or nomination.
    # If after the the configured number, the candidate is yet to answer a binding request or a nomination we set the pair as failed.
    max_binding_requests: 7

    # SetNAT1To1IPs sets a list of external IP addresses of 1:1 (D)NAT and a candidate type for which the external IP address is used.
    # This is useful when you are host a server using Pion on an AWS EC2 instance which has a private address, behind a 1:1 DNAT with a public IP (e.g. Elastic IP).
    # In this case, you can give the public IP address so that Pion will use the public IP address in its candidate instead of the private IP address.
    nat_1to1_ips: []

    # Limit the port range used by ICE
    port:
        min: 49152
        max: 65535

    # The check interval controls how often our task loop runs when in the connecting state.
    check_interval: 200ms
    
    # If the duration is 0, the ICE Agent will never go to disconnected
    disconnected_timeout: 5s

    # If the duration is 0, we will never go to failed.
    failed_timeout: 5s
    restart_timeout: 5s

    # Determines how often should we send ICE keepalives (should be less then connection timeout above).
    # A keepalive interval of 0 means we never send keepalive packets
    keepalive_interval: 2s
```

## Environment Variables

All the settings from the configuration file can also be passed via environment variables by following the following rules:

-   Convert the setting name to uppercase
-   Prefixing the setting name with `CUNICU_`
-   Nested settings are separated by underscores

**Example:** The setting `endpoint_disc.ice.max_binding_requests` can be set by the environment variable `CUNICU_ENDPOINT_DISC_ICE_MAX_BINDING_REQUESTS`

**Note:** Setting lists such as `endpoint_disc.ice.urls` or `backends` can currently not be set via environment variables.

## DNS Auto-configuration

cunīcu als supports retrieving parts of the configuration via DNS lookups.

When `cunicu daemon` is started with a `--domain example.com` parameter it will look for the following DNS records to obtain its configuration.

STUN and TURN servers used for ICE are retrieved by SVR lookups and other cunīcu settings are retrieved via TXT lookups: 

```text
_stun._udp.example.com.  3600 IN SRV 10 0 3478 stun.example.com.
_stuns._tcp.example.com. 3600 IN SRV 10 0 3478 stun.example.com.
_turn._udp.example.com.  3600 IN SRV 10 0 3478 turn.example.com.
_turn._tcp.example.com.  3600 IN SRV 10 0 3478 turn.example.com.
_turns._tcp.example.com. 3600 IN SRV 10 0 5349 turn.example.com.

example.com.             3600 IN TXT "cunicu-backend=p2p"
example.com.             3600 IN TXT "cunicu-peer-disc-community=my-community-password"
example.com.             3600 IN TXT "cunicu-endpoint-disc-ice-username=user1"
example.com.             3600 IN TXT "cunicu-endpoint-disc-ice-password=pass1"
example.com.             3600 IN TXT "cunicu-config=https://example.com/cunicu.yaml"
```

**Note:** The `cunicu-backend` and `cunicu-config` TXT records can be provided multiple times. Others not.

## Remote Configuration File

When `cunicu daemon` can be started with `--config` options pointing to HTTPS URIs.
cunīcu will download all configuration files in the order they are specified on the command line and merge them subsequently.

This feature can be combined with the DNS auto-configuration method by providing a TXT record pointing to the configuration file:

```text
example.com.             3600 IN TXT "cunicu-config=https://example.com/cunicu.yaml"
```

**Note:** Remote configuration files must be fetched via HTTPS if they are not hosted locally and required a trusted server certificate.

pkg/args/args.go
package args

import (
	"flag"
	"fmt"
	"io"
	"net/url"
	"os"
	"regexp"
	"strconv"
	"strings"
	"time"

	"go.uber.org/zap"
	"go.uber.org/zap/zapcore"
	pice "riasc.eu/wice/internal/ice"
	"riasc.eu/wice/pkg/proxy"
	"riasc.eu/wice/pkg/signaling"

	"github.com/pion/ice/v2"
)

// Copied from pion/ice/agent_config.go
const (
	// defaultCheckInterval is the interval at which the agent performs candidate checks in the connecting phase
	defaultCheckInterval = 200 * time.Millisecond

	// keepaliveInterval used to keep candidates alive
	defaultKeepaliveInterval = 2 * time.Second

	// defaultDisconnectedTimeout is the default time till an Agent transitions disconnected
	defaultDisconnectedTimeout = 5 * time.Second

	// defaultFailedTimeout is the default time till an Agent transitions to failed after disconnected
	defaultFailedTimeout = 25 * time.Second

	// max binding request before considering a pair failed
	defaultMaxBindingRequests = 7
)

var (
	defaultICEUrls = []*ice.URL{
		{
			Scheme:   ice.SchemeTypeSTUN,
			Host:     "stun.l.google.com",
			Port:     19302,
			Username: "",
			Password: "",
			Proto:    ice.ProtoTypeUDP,
		},
	}
)

type backendURLList []*url.URL

func (i *backendURLList) String() string {
	s := []string{}
	for _, u := range *i {
		s = append(s, u.String())
	}

	return strings.Join(s, ",")
}

func (i *backendURLList) Set(value string) error {

	// Allow the user to specify just the backend type as a valid url.
	// E.g. "p2p" instead of "p2p:"
	if !strings.Contains(value, ":") {
		value += ":"
	}

	uri, err := url.Parse(value)
	if err != nil {
		return fmt.Errorf("invalid backend URI: %w", err)
	}

	*i = append(*i, uri)

	return nil
}

type iceURLList []*ice.URL

func (i *iceURLList) String() string {
	s := []string{}
	for _, u := range *i {
		s = append(s, u.String())
	}

	return strings.Join(s, ",")
}

func (i *iceURLList) Set(value string) error {
	iceUrl, err := ice.ParseURL(value)
	if err != nil {
		return fmt.Errorf("failed to parse ICE url %s: %w", value, err)
	}

	*i = append(*i, iceUrl)

	return nil
}

type arrayFlags []string

func (i *arrayFlags) String() string {
	return strings.Join(*i, ",")
}

func (i *arrayFlags) Set(value string) error {
	*i = append(*i, value)
	return nil
}

type Args struct {
	Backends        []*url.URL
	User            bool
	ProxyType       proxy.Type
	ConfigSync      bool
	ConfigPath      string
	WatchInterval   time.Duration
	RestartInterval time.Duration

	InterfaceRegex    *regexp.Regexp
	IceInterfaceRegex *regexp.Regexp
	AgentConfig       ice.AgentConfig

	Socket     string
	SocketWait bool

	Interfaces []string
}

func showUsage() {
	fmt.Fprintf(os.Stderr, "usage: %s [OPTIONS] [IFACES ...]\n", os.Args[0])
	fmt.Println()
	fmt.Println("  IFACES  is a list of Wireguard interfaces")
	fmt.Println("          (defaults to all available Wireguard interfaces)")
	fmt.Println("")
	fmt.Println(("Available OPTIONS are:"))
	flag.PrintDefaults()
	fmt.Println()
	fmt.Println("  (**) These options can be specified multiple times")
	fmt.Println()
	fmt.Println("Available backends types are:")
	for name, plugin := range signaling.Backends {
		fmt.Printf("  %-7s %s\n", name, plugin.Description)
	}
}

func (a *Args) DumpConfig(wr io.Writer) {
	fmt.Fprintln(wr, "Options:")
	fmt.Fprintln(wr, "  URLs:")
	for _, u := range a.AgentConfig.Urls {
		fmt.Fprintf(wr, "    %s\n", u.String())
	}

	fmt.Fprintln(wr, "  Interfaces:")
	for _, d := range a.Interfaces {
		fmt.Fprintf(wr, "    %s\n", d)
	}

	fmt.Fprintf(wr, "  User: %s\n", strconv.FormatBool(a.User))
	fmt.Fprintf(wr, "  ProxyType: %s\n", a.ProxyType.String())

	fmt.Fprintf(wr, "  Signalling Backends:\n")
	for _, b := range a.Backends {
		fmt.Fprintf(wr, "    %s\n", b)
	}
}

func candidateTypeFromString(t string) (ice.CandidateType, error) {
	switch t {
	case "host":
		return ice.CandidateTypeHost, nil
	case "srflx":
		return ice.CandidateTypeServerReflexive, nil
	case "prflx":
		return ice.CandidateTypePeerReflexive, nil
	case "relay":
		return ice.CandidateTypeRelay, nil
	default:
		return ice.CandidateTypeUnspecified, fmt.Errorf("unknown candidate type: %s", t)
	}
}

func networkTypeFromString(t string) (ice.NetworkType, error) {
	switch t {
	case "udp4":
		return ice.NetworkTypeUDP4, nil
	case "udp6":
		return ice.NetworkTypeUDP6, nil
	case "tcp4":
		return ice.NetworkTypeTCP4, nil
	case "tcp6":
		return ice.NetworkTypeTCP6, nil
	default:
		return ice.NetworkTypeTCP4, fmt.Errorf("unknown network type: %s", t)
	}
}

func Parse(progname string, argv []string) (*Args, error) {
	var err error
	var iceURLs, iceCandidateTypes, iceNetworkTypes, iceNat1to1IPs arrayFlags
	var backendURLs backendURLList
	var logLevel zapcore.Level

	flags := flag.NewFlagSet(progname, flag.ContinueOnError)

	flags.Usage = showUsage

	flags.Var(&logLevel, "log-level", "log level (one of \"panic\", \"fatal\", \"error\", \"warn\", \"info\", \"debug\", \"trace\")")
	flags.Var(&backendURLs, "backend", "backend type / URL")
	user := flags.Bool("user", false, "start userspace Wireguard daemon")
	proxyType := flags.String("proxy", "auto", "proxy type to use")
	interfaceFilter := flags.String("interface-filter", ".*", "regex for filtering Wireguard interfaces (e.g. \"wg-.*\")")
	configSync := flags.Bool("config-sync", false, "sync Wireguard interface with configuration file (see \"wg synconf\"")
	configPath := flags.String("config-path", "/etc/wireguard", "base path to search for Wireguard configuration files")
	watchInterval := flags.Duration("watch-interval", time.Second, "interval at which we are polling the kernel for updates on the Wireguard interfaces")

	// ice.AgentConfig fields
	flags.Var(&iceURLs, "url", "STUN and/or TURN server address  (**)")
	flags.Var(&iceCandidateTypes, "ice-candidate-type", "usable candidate types (**, one of \"host\", \"srflx\", \"prflx\", \"relay\")")
	flags.Var(&iceNetworkTypes, "ice-network-type", "usable network types (**, select from \"udp4\", \"udp6\", \"tcp4\", \"tcp6\")")
	flags.Var(&iceNat1to1IPs, "ice-nat-1to1-ip", "list of IP addresses which will be added as local server reflexive candidates (**)")

	icePortMin := flags.Uint("ice-port-min", 0, "minimum port for allocation policy (range: 0-65535)")
	icePortMax := flags.Uint("ice-port-max", 0, "maximum port for allocation policy (range: 0-65535)")
	iceLite := flags.Bool("ice-lite", false, "lite agents do not perform connectivity check and only provide host candidates")
	iceMdns := flags.Bool("ice-mdns", false, "enable local Multicast DNS discovery")
	iceMaxBindingRequests := flags.Int("ice-max-binding-requests", defaultMaxBindingRequests, "maximum number of binding request before considering a pair failed")
	iceInsecureSkipVerify := flags.Bool("ice-insecure-skip-verify", false, "skip verification of TLS certificates for secure STUN/TURN servers")
	iceInterfaceFilter := flags.String("ice-interface-filter", ".*", "regex for filtering local interfaces for ICE candidate gathering (e.g. \"eth[0-9]+\")")
	iceDisconnectedTimeout := flags.Duration("ice-disconnected-timout", defaultDisconnectedTimeout, "time till an Agent transitions disconnected")
	iceFailedTimeout := flags.Duration("ice-failed-timeout", defaultFailedTimeout, "time until an Agent transitions to failed after disconnected")
	iceKeepaliveInterval := flags.Duration("ice-keepalive-interval", defaultKeepaliveInterval, "interval netween STUN keepalives")
	iceCheckInterval := flags.Duration("ice-check-interval", defaultCheckInterval, "interval at which the agent performs candidate checks in the connecting phase")
	iceRestartInterval := flags.Duration("ice-restart-interval", defaultDisconnectedTimeout, "time to wait before ICE restart")
	iceUsername := flags.String("ice-user", "", "username for STUN/TURN credentials")
	icePassword := flags.String("ice-pass", "", "password for STUN/TURN credentials")
	// iceMaxBindingRequestTimeout := flag.Duration("ice-max-binding-request-timeout", maxBindingRequestTimeout, "wait time before binding requests can be deleted")

	socket := flags.String("socket", "/var/run/wice.sock", "Unix control and monitoring socket")
	socketWait := flags.Bool("socket-wait", false, "wait until first client connected to control socket before continuing start")

	if err := flags.Parse(argv); err != nil {
		return nil, fmt.Errorf("failed to parse args: %w", err)
	}

	logger := zap.L().Named("args")

	logger.Info("Setting debug level", zap.Any("level", logLevel))
	// logger.SetLevel(logLevel.Level)

	args := &Args{
		User:      *user,
		Backends:  backendURLs,
		ProxyType: proxy.ProxyTypeFromString(*proxyType),
		// Discover:        *discover,
		ConfigSync:      *configSync,
		ConfigPath:      *configPath,
		WatchInterval:   *watchInterval,
		RestartInterval: *iceRestartInterval,
		Socket:          *socket,
		SocketWait:      *socketWait,
		Interfaces:      flag.Args(),
		AgentConfig: ice.AgentConfig{
			PortMin:            uint16(*icePortMin),
			PortMax:            uint16(*icePortMax),
			Lite:               *iceLite,
			InsecureSkipVerify: *iceInsecureSkipVerify,
			Urls:               []*ice.URL{},
		},
	}

	// Find best proxy method
	if args.ProxyType == proxy.TypeAuto {
		args.ProxyType = proxy.AutoProxy()
	}

	// Check proxy type
	if args.ProxyType == proxy.TypeInvalid {
		return nil, fmt.Errorf("invalid proxy type: %s", *proxyType)
	}

	// Compile interface regex
	args.InterfaceRegex, err = regexp.Compile(*interfaceFilter)
	if err != nil {
		return nil, fmt.Errorf("invalid interface filter: %w", err)
	}

	if *iceMaxBindingRequests >= 0 {
		maxBindingReqs := uint16(*iceMaxBindingRequests)
		args.AgentConfig.MaxBindingRequests = &maxBindingReqs
	}
	if *iceMdns {
		args.AgentConfig.MulticastDNSMode = ice.MulticastDNSModeQueryAndGather
	}
	if *iceDisconnectedTimeout > 0 {
		args.AgentConfig.DisconnectedTimeout = iceDisconnectedTimeout
	}
	if *iceFailedTimeout > 0 {
		args.AgentConfig.FailedTimeout = iceFailedTimeout
	}
	if *iceKeepaliveInterval > 0 {
		args.AgentConfig.KeepaliveInterval = iceKeepaliveInterval
	}
	if *iceCheckInterval > 0 {
		args.AgentConfig.CheckInterval = iceCheckInterval
	}
	if len(iceNat1to1IPs) > 0 {
		args.AgentConfig.NAT1To1IPCandidateType = ice.CandidateTypeServerReflexive
		args.AgentConfig.NAT1To1IPs = iceNat1to1IPs
	}

	args.IceInterfaceRegex, err = regexp.Compile(*iceInterfaceFilter)
	if err != nil {
		return nil, fmt.Errorf("failed to compile interface regex: %w", err)
	}

	// Parse candidate types
	for _, c := range iceCandidateTypes {
		candType, err := candidateTypeFromString(c)
		if err != nil {
			return nil, err
		}
		args.AgentConfig.CandidateTypes = append(args.AgentConfig.CandidateTypes, candType)
	}

	// Parse network types
	if len(iceNetworkTypes) == 0 {
		args.AgentConfig.NetworkTypes = []ice.NetworkType{
			ice.NetworkTypeUDP4,
			ice.NetworkTypeUDP6,
		}
	} else {
		for _, n := range iceNetworkTypes {
			netType, err := networkTypeFromString(n)
			if err != nil {
				return nil, err
			}
			args.AgentConfig.NetworkTypes = append(args.AgentConfig.NetworkTypes, netType)
		}
	}

	// Add default backend
	if len(args.Backends) == 0 {
		args.Backends = append(args.Backends, &url.URL{
			Scheme: "p2p",
		})
	}

	// Add default STUN/TURN servers
	if len(args.AgentConfig.Urls) == 0 {
		args.AgentConfig.Urls = defaultICEUrls
	} else {
		// Set ICE credentials
		for _, u := range args.AgentConfig.Urls {
			if *iceUsername != "" {
				u.Username = *iceUsername
			}

			if *icePassword != "" {
				u.Password = *icePassword
			}
		}
	}

	args.AgentConfig.LoggerFactory = &pice.LoggerFactory{}

	return args, nil
}
